### PR TITLE
[HTTP/3] Try catch for unreachable host in AltSvc H/3 upgrade test

### LINKED ISSUE #70971: Test failure net7.0-windows-Release-x64-CoreCLR_release-Windows.Amd64.Server2022.Open System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success
Run: [runtime-libraries-coreclr outerloop 20220619.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1833104&view=ms.vss-test-web.build-test-results-tab&runId=48491140&resultId=104547&paneView=debug)

Failed test:
```
net7.0-windows-Release-x64-CoreCLR_release-Windows.Amd64.Server2022.Open

- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success(uri: \"https://quic.aiortc.org:443/\")

net7.0-windows-Release-x64-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open

- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success(uri: \"https://quic.aiortc.org:443/\")

net7.0-windows-Release-x64-CoreCLR_release-Windows.11.Amd64.Client.Open

- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success(uri: \"https://quic.aiortc.org:443/\")

net7.0-windows-Release-x86-CoreCLR_release-Windows.11.Amd64.Client.Open

- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success(uri: \"https://quic.aiortc.org:443/\")

net7.0-windows-Release-x86-CoreCLR_release-Windows.Amd64.Server2022.Open

- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success(uri: \"https://quic.aiortc.org:443/\")

net7.0-windows-Release-x86-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open

- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http3_MsQuic.Public_Interop_Upgrade_Success(uri: \"https://quic.aiortc.org:443/\")

```
**Error message:**
```
System.Net.Http.HttpRequestException : No connection could be made because the target machine actively refused it. (quic.aiortc.org:443)
---- System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it.


Stack trace
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 1486
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 1425
   at System.Net.Http.HttpConnectionPool.AddHttp2ConnectionAsync(QueueItem queueItem) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 708
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken) in /_/src/libraries/Common/src/System/Threading/Tasks/TaskCompletionSourceWithCancellation.cs:line 23
   at System.Net.Http.HttpConnectionPool.GetHttp2ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 823
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 1102
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs:line 85
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 558
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http3.Public_Interop_Upgrade_Success(String uri) in /_/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:line 642
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:line 1365
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|281_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs:line 126
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 1486
```

### PR DESCRIPTION
**Pull request by ManickaP**
Fixes https://github.com/dotnet/runtime/issues/70971

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs [hunk 1]
```diff
@@ -639,9 +640,17 @@ public async Task Public_Interop_Upgrade_Success(string uri)
...
-                Assert.NotEqual(3, responseA.Version.Major);
+                try
+                {
+                    using HttpResponseMessage responseA = await client.SendAsync(requestA).WaitAsync(TimeSpan.FromSeconds(20));
+                    Assert.Equal(HttpStatusCode.OK, responseA.StatusCode);
+                    Assert.NotEqual(3, responseA.Version.Major);
+                }
+                catch (HttpRequestException ex) when (ex.InnerException is SocketException se && se.SocketErrorCode == SocketError.NetworkUnreachable)
+                {
+                    _output.WriteLine($"Unable to establish non-H/3 connection to {uri}: {ex}");
```
Comment (rzikm) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:651]: Is the output written to stdout if the test completes successfully? I thought it happens only on errors.

Should we perhaps use SkipTestException?

Comment (ManickaP) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:651]: This is written to test output and stored in testResults.xml, you can see it even in success.

**Review by rzikm — APPROVED** [review]

**Review by CarnaViire — APPROVED** [review]

**Comment by ManickaP** [comment]
> Would it make sense to change the version policy to require exact match since this is H3 test?

We have two different tests for H/3 interop, one testing exact version: 
https://github.com/dotnet/runtime/blob/f9999cc1aaa2be9fdc81e69f78c3771e6106d93c/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs#L559

And this one, testing Alt-Svc upgrade. So we do 2 requests, one for H/1/2 and following for H/3 on the same client. What failed here is the first request since that particular endpoint stopped supporting lower than H/3 versions. And since we're using the same endpoint list for both tests, I kept the endpoint there but handled it with a catch block.

**Comment by ManickaP** [comment]
Failures are: https://github.com/dotnet/runtime/issues/71037

**Pull request merged**